src/libtrx/game/rooms/common.c

```diff
@@ -21,3 +43,127 @@ int32_t Room_GetAdjoiningRooms(
+            break;
+
+        case FT_ROOF:
+            sector->ceiling.tilt = *data++;
+            break;
+
+        case FT_DOOR:
+            sector->portal_room.wall = *data++;
+            break;
+
+        case FT_LAVA:
+            sector->is_death_sector = true;
+            break;
+
+#if TR_VERSION == 2
+        case FT_CLIMB:
+            sector->ladder = (LADDER_DIRECTION)FD_LADDER_TYPE(fd_entry);
+            break;
+#endif
+
+        case FT_TRIGGER: {
+            assert(sector->trigger == NULL);
+
+            TRIGGER *const trigger =
+                GameBuf_Alloc(sizeof(TRIGGER), GBUF_FLOOR_DATA);
+            sector->trigger = trigger;
+
+            const int16_t trig_setup = *data++;
+            trigger->type = FD_TRIG_TYPE(fd_entry);
+            trigger->timer = FD_TRIG_TIMER(trig_setup);
+            trigger->one_shot = FD_TRIG_ONE_SHOT(trig_setup);
+            trigger->mask = FD_TRIG_MASK(trig_setup);
+            trigger->item_index = NO_ITEM;
+            trigger->command_count = 0;
+
+            if (trigger->type == TT_SWITCH || trigger->type == TT_KEY
+                || trigger->type == TT_PICKUP) {
+                const int16_t item_data = *data++;
+                trigger->item_index = FD_TRIG_CMD_ARG(item_data);
+                if (FD_IS_DONE(item_data)) {
+                    break;
+                }
+            }
+
+            const int16_t *command_data = data;
+            while (true) {
+                trigger->command_count++;
+                int16_t command = *data++;
+                if (FD_TRIG_CMD_TYPE(command) == TO_CAMERA) {
+                    command = *data++;
+                }
+                if (FD_IS_DONE(command)) {
+                    break;
+                }
+            }
+
+            trigger->commands = GameBuf_Alloc(
+                sizeof(TRIGGER_CMD) * trigger->command_count, GBUF_FLOOR_DATA);
+            for (int32_t i = 0; i < trigger->command_count; i++) {
+                int16_t command = *command_data++;
+                TRIGGER_CMD *const cmd = &trigger->commands[i];
+                cmd->type = FD_TRIG_CMD_TYPE(command);
+
+                if (cmd->type == TO_CAMERA) {
+                    TRIGGER_CAMERA_DATA *const cam_data = GameBuf_Alloc(
+                        sizeof(TRIGGER_CAMERA_DATA), GBUF_FLOOR_DATA);
+                    cmd->parameter = (void *)cam_data;
+                    cam_data->camera_num = FD_TRIG_CMD_ARG(command);
+
+                    command = *command_data++;
+                    cam_data->timer = FD_TRIG_TIMER(command);
+                    cam_data->glide = FD_TRIG_CAM_GLIDE(command);
+                    cam_data->one_shot = FD_TRIG_ONE_SHOT(command);
+                } else {
+                    cmd->parameter = (void *)(intptr_t)FD_TRIG_CMD_ARG(command);
+                }
+            }
+
+            break;
+        }
+
+        default:
+            break;
+        }
+    } while (!FD_IS_DONE(fd_entry));
+}
```

Comment:
If we're going to get significantly more data related to floor data triggers, maybe we could make its own file for it `game/rooms/floor_data.[ch]`?

Reply:
Moved trigger reading to its own function, the main loop is more readable now IMO.